include/seastar/core/shared_future.hh

```diff
@@ -141,41 +145,47 @@ private:
             // Don't warn if the shared future is exceptional. Any
             // warnings will be reported by the futures returned by
             // get_future.
-            if (_original_future.failed()) {
-                _original_future.ignore_ready_future();
+            if (this->_state.failed()) {
+                this->_state.ignore();
             }
         }
-        explicit shared_state(future_type f) noexcept : _original_future(std::move(f)) { }
-        void resolve(future_type&& f) noexcept {
-            _original_future = std::move(f);
-            auto& state = _original_future._state;
-            if (_original_future.failed()) {
+        explicit shared_state(future_type f) noexcept
+                : _keepaliver(f.available() ? nullptr : this->shared_from_this()) {
+            if (f.available()) {
+                this->set_state(f.get_available_state_ref());
+            } else {
+                internal::set_callback(std::move(f), this);
+            }
+        }
+        virtual void run_and_dispose() noexcept override {
+            if (this->_state.failed()) {
                 while (_peers) {
-                    _peers.front().pr.set_exception(state.get_exception());
+                    _peers.front().pr.set_exception(this->_state.get_exception());
                     _peers.pop_front();
                 }
             } else {
                 while (_peers) {
                     auto& p = _peers.front().pr;
                     try {
-                        p.set_value(state.get_value());
+                        p.set_value(this->_state.get_value());
                     } catch (...) {
                         p.set_exception(std::current_exception());
                     }
                     _peers.pop_front();
                 }
             }
+            _keepaliver.release();
```

Comment:
This is nice

Reply:
what if the `shared_state` is never scheduled? it would lead to leak of it.

Reply:
#1937 was created to track this regression.

Reply:
Why wouldn't it be scheduled? It is scheduled here, only if the result of the future is not available yet:

```c++
        explicit shared_state(future_type f) noexcept
                : _keepaliver(f.available() ? nullptr : this->shared_from_this()) {
            if (f.available()) {
                this->set_state(f.get_available_state_ref());
            } else {
                internal::set_callback(std::move(f), this);
            }
        }
```

How is it different from e.g. a continuation created by `.then()`?

Reply:
@piodul hi Piotr, the "not-scheduled" case is but a hypothesis. what i really want to address is #1937. maybe, just maybe, the corresponding promise is never fulfilled?